File: frontend/src/utils/memoryGuard.js
/**
 * Garde-fou mémoire pour le frontend Vue.js
 */

class MemoryGuard {
  constructor(maxMemoryMB = 100, checkIntervalMs = 5000) {
    this.maxMemoryBytes = maxMemoryMB * 1024 * 1024
    this.checkInterval = checkIntervalMs
    this.warningThreshold = this.maxMemoryBytes * 0.8 // 80% = warning
    this.intervalId = null
    this.isMonitoring = false
  }

  /**
   * Obtient l'utilisation mémoire actuelle (si disponible)
   */
  getMemoryUsage() {
    if (performance.memory) {
      return {
        used: performance.memory.usedJSHeapSize,
        total: performance.memory.totalJSHeapSize,
        limit: performance.memory.jsHeapSizeLimit
      }
    }
    return null
  }

  /**
   * Vérifie si la limite mémoire est dépassée
   */
  checkMemoryLimit() {
    const memory = this.getMemoryUsage()
    if (!memory) {
      return null // API non disponible dans ce navigateur
    }

    const currentMB = memory.used / (1024 * 1024)
    const maxMB = this.maxMemoryBytes / (1024 * 1024)

    if (memory.used > this.maxMemoryBytes) {
      return `LIMITE MÉMOIRE DÉPASSÉE: ${currentMB.toFixed(1)}MB > ${maxMB}MB`
    } else if (memory.used > this.warningThreshold) {
      const percentage = (memory.used / this.maxMemoryBytes * 100).toFixed(1)
      return `AVERTISSEMENT MÉMOIRE: ${currentMB.toFixed(1)}MB / ${maxMB}MB (${percentage}%)`
    }

    return null
  }

  /**
   * Force le nettoyage et l'arrêt de l'application
   */
  forceShutdown(reason) {
    console.error(`🚨 ARRÊT D'URGENCE: ${reason}`)
    const memory = this.getMemoryUsage()
    if (memory) {
      console.error(`Mémoire actuelle: ${(memory.used / (1024 * 1024)).toFixed(1)}MB`)
    }
    console.error('Application fermée pour éviter un crash navigateur')

    // Nettoyer les ressources
    this.stopMonitoring()

    // Forcer le garbage collector si possible
    if (window.gc) {
      try {
        window.gc()
      } catch (e) {
        // Ignoré
      }
    }

    // Afficher un message à l'utilisateur
    alert('⚠️ Limite mémoire dépassée - Application fermée pour votre sécurité')

    // Rediriger vers une page d'erreur ou recharger
    window.location.reload()
  }

  /**
   * Démarre la surveillance mémoire
   */
  startMonitoring() {
    if (this.isMonitoring) return

    this.isMonitoring = true
    this.intervalId = setInterval(() => {
      const memoryCheck = this.checkMemoryLimit()
      if (memoryCheck) {
        if (memoryCheck.includes('LIMITE MÉMOIRE DÉPASSÉE')) {
          this.forceShutdown(memoryCheck)
        } else {
          console.warn(`⚠️ ${memoryCheck}`)
        }
      }
    }, this.checkInterval)

    console.log(`🛡️ Surveillance mémoire activée (limite: ${this.maxMemoryBytes / (1024 * 1024)}MB)`)
  }

  /**
   * Arrête la surveillance mémoire
   */
  stopMonitoring() {
    if (this.intervalId) {
      clearInterval(this.intervalId)
      this.intervalId = null
    }
    this.isMonitoring = false
    console.log('🛡️ Surveillance mémoire désactivée')
  }

  /**
   * Obtient des statistiques mémoire formatées
   */
  getMemoryStats() {
    const memory = this.getMemoryUsage()
    if (!memory) {
      return 'Statistiques mémoire non disponibles dans ce navigateur'
    }

    return {
      used: `${(memory.used / (1024 * 1024)).toFixed(1)} MB`,
      total: `${(memory.total / (1024 * 1024)).toFixed(1)} MB`,
      limit: `${(memory.limit / (1024 * 1024)).toFixed(1)} MB`,
      percentage: `${((memory.used / memory.limit) * 100).toFixed(1)}%`
    }
  }
}

// Instance globale
export const memoryGuard = new MemoryGuard(100, 5000) // 100MB, vérification toutes les 5s

// Démarrer automatiquement la surveillance
if (typeof window !== 'undefined') {
  memoryGuard.startMonitoring()
}

export default MemoryGuard

File: src/back_office_lmelp/utils/memory_guard.py
"""Garde-fou mémoire pour éviter les fuites."""

import psutil


class MemoryGuard:
    """Surveille et limite l'utilisation mémoire."""

    def __init__(self, max_memory_mb: int = 1000, check_interval: int = 10):
        """
        Initialise le garde-fou mémoire.

        Args:
            max_memory_mb: Limite mémoire en MB (défaut: 1000MB)
            check_interval: Intervalle de vérification en secondes
        """
        self.max_memory_bytes = max_memory_mb * 1024 * 1024
        self.check_interval = check_interval
        self.process = psutil.Process()
        self.warning_threshold = self.max_memory_bytes * 0.8  # 80% = warning

    def get_memory_usage(self) -> int:
        """Retourne l'utilisation mémoire actuelle en bytes."""
        return int(self.process.memory_info().rss)

    def get_memory_usage_mb(self) -> float:
        """Retourne l'utilisation mémoire actuelle en MB."""
        return self.get_memory_usage() / (1024 * 1024)

    def check_memory_limit(self) -> str | None:
        """
        Vérifie si la limite mémoire est dépassée.

        Returns:
            Message d'erreur si limite dépassée, None sinon
        """
        current_memory = self.get_memory_usage()
        current_mb = current_memory / (1024 * 1024)
        max_mb = self.max_memory_bytes / (1024 * 1024)

        if current_memory > self.max_memory_bytes:
            return f"LIMITE MÉMOIRE DÉPASSÉE: {current_mb:.1f}MB > {max_mb}MB"
        if current_memory > self.warning_threshold:
            return f"AVERTISSEMENT MÉMOIRE: {current_mb:.1f}MB / {max_mb}MB ({current_mb / max_mb * 100:.1f}%)"

        return None

    def force_shutdown(self, reason: str) -> None:
        """Force l'arrêt de l'application de manière plus gracieuse."""
        print(f"🚨 ARRÊT D'URGENCE: {reason}")
        print(f"Mémoire actuelle: {self.get_memory_usage_mb():.1f}MB")
        print("Application fermée pour éviter un crash système")

        # Nettoyer les ressources
        try:
            # Forcer le garbage collector
            import gc

            gc.collect()
            print("🧹 Garbage collector exécuté")
        except Exception:
            pass

        # Essayer d'arrêter le serveur proprement si possible
        try:
            # Importer ici pour éviter les dépendances circulaires
            from back_office_lmelp.app import _server_instance

            if _server_instance is not None:
                _server_instance.should_exit = True
                print("📡 Signal d'arrêt envoyé au serveur")
                # Donner un peu de temps pour l'arrêt gracieux
                import time

                time.sleep(0.5)
        except ImportError:
            # Si on ne peut pas importer (par exemple dans les tests),
            # on continue avec l'arrêt normal
            pass
        except Exception as e:
            print(f"⚠️ Erreur lors de l'arrêt gracieux: {e}")

        # Arrêt via exception plutôt qu'os._exit pour permettre le cleanup
        raise SystemExit(1)


# Instance globale
memory_guard = MemoryGuard(max_memory_mb=1000)
